### PR TITLE
reclaim locks

### PR DESCRIPTION
**Pull request by artushin**
This is an alternative to fraenkel's solution for Issue #855 as discussed here: https://github.com/hashicorp/consul/pull/861

### CODE REVIEW

**Comment by artushin** [comment]
There was a bug which would result in an invalid session with the prior code if lock struct was reused (Lock(), Unlock(), Lock()). Should work now and added a test around the problem.

File: api/lock.go [hunk 1]
```diff
@@ -123,23 +123,12 @@ func (l *Lock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
 	}
 
 	// Check if we need to create a session first
-	l.lockSession = l.opts.Session
```
Comment (armon) [inline api/lock.go]: Removing this means that the lock can't be parameterized with a session any longer

File: api/lock.go [hunk 2]
```diff
@@ -165,25 +154,36 @@ WAIT:
...
 	}
 
-	// Try to acquire the lock
-	lockEnt := l.lockEntry(l.lockSession)
-	locked, _, err := kv.Acquire(lockEnt, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to acquire lock: %v", err)
-	}
+	if pair == nil || pair.Session == "" {
+		if l.lockSession != "" {
```
Comment (armon) [inline api/lock.go]: Why do we need this logic again? We should have established the lockSession at this point

File: api/lock.go [hunk 3]
```diff
@@ -288,9 +288,19 @@ func (l *Lock) createSession() (string, error) {
...
+		return nil, err
 	}
-	return id, nil
+
+	l.sessionRenew = make(chan struct{})
+	l.lockSession = id
+	go session.RenewPeriodic(l.opts.SessionTTL, id, nil, l.sessionRenew)
+
+	return func() {
+		if !l.isHeld {
```
Comment (armon) [inline api/lock.go]: Its not safe to manipulate isHeld without holding the proper locks

File: api/lock.go [hunk 4]
```diff
@@ -306,7 +316,10 @@ func (l *Lock) lockEntry(session string) *KVPair {
 // monitorLock is a long running routine to monitor a lock ownership
 // It closes the stopCh if we lose our leadership.
 func (l *Lock) monitorLock(session string, stopCh chan struct{}) {
-	defer close(stopCh)
+	defer func() {
+		l.isHeld = false
```
Comment (armon) [inline api/lock.go]: Also unsafe here

**Comment by artushin** [comment]
Hey Armon, sorry, I was on vacation for a bit and didn't get a change to look at this. I'll take a look now.

File: api/lock.go (hunk 1 above)
Comment (artushin) [inline api/lock.go]: Not sure what you mean here. The SessionName and SessionTTL would still coming from the opts. It's the id created by createSession that would get updated. Maybe I'm not seeing a use case in which you would want that to happen outside of the Lock(), Unlock(), Destroy() API but it seems to me that it can all be handled internally.

File: api/lock.go (hunk 2 above)
Comment (artushin) [inline api/lock.go]: This is for if the actual lock ends up being unlocked but the session is now invalid.

File: api/lock.go (hunk 3 above)
Comment (artushin) [inline api/lock.go]: The returned little def function here is run within the Lock, Unlock, Destroy functions after the defer Unlock(), and since defer is LIFO, it should be safe.

File: api/lock.go (hunk 4 above)
Comment (artushin) [inline api/lock.go]: Yep, that's an issue. Will fix.

File: api/lock.go (hunk 2 above)
Comment (artushin) [inline api/lock.go]: Just to further clarify this, let's say entity A acquires the lock. If it then loses the lock (and monitorLock notifies it of such), it will try to reacquire the lock by calling .Lock() again. Since monitorLock does not reset the l.lockSession, we still have the same lockSession and will try to reclaim the lock with that session. The server responds with an invalid session error in this case. I think there's an argument for doing all of the session work outside of the api (and that goes back to the previous comment about using a session ID from opts) but if we're already calling createSession internally, I think we should be recreating a valid session for the user in this case.

Comment (artushin) [inline api/lock.go]: Wait, I think I see what you're talking about now. That logic doesn't make sense because we always have a l.lockSession != nil. Hmm. I remember there was an invalid session issue. Let me look further.

**Comment by artushin** [comment]
Ok, did a bit of a refactor which should address all of the comments. Still remember there being an invalid session issue at one point but can't replicate it now. When you're taking a look at this, can you keep that in mind and let me know if you see any places where we could end up with an invalid session?

File: api/lock.go [hunk 5]
```diff
@@ -99,7 +99,9 @@ func (c *Client) LockOpts(opts *LockOptions) (*Lock, error) {
 	l := &Lock{
 		c:    c,
 		opts: opts,
+		held: make(chan struct{}, 1),
```
Comment (armon) [inline api/lock.go]: I don't see why this change is necessary?

**Comment by armon** [comment]
I think there is a deeper issue to consider about the semantics of Lock(). If you pass in a session via `LockOpts.Session` then the lifecycle management of the session is assumed external to Lock. If that is blank, then the Lock creates a session and the lifecycle of the session is tied to `Lock()` and `Unlock()`.

I think understanding those semantics is critical, as this changes those. Firstly, this change makes it impossible to provide a session as far as I can tell. Secondly, the lifecycle of the created session seems to be undefined with this change. The session is created on the first call to Lock() but when is it expected to be released?

My view is that if you want to re-try acquisition, you should manage the session lifecycle outside of Lock.

**Comment by artushin** [comment]
I just pushed a unit test for the being able to provide a session which is managed externally (I added back the line which set lockSession from opts in the previous commit). As far as the the lifecycle of the created session goes, the session is only invalided on explicitly calling Unlock. I think that semantically, that actually makes sense. If I create a new Lock, I would be attempting to reclaim the lock by calling Lock() on it if the lock is lost, or invalidating the session with Unlock(), and then either setting up a new session by calling Lock() again, or creating a new Lock with a new session. In the latter case, I think the semantics make sense because the only way to pass a session in would be to create a new Lock.
